### ISSUE TITLE
Jackson ignores `@JsonFormat` on field if the object is built via constructor (uses `@JsonCreator`)

### ISSUE THREAD

**Issue by bric3** [FasterXML/jackson-databind#2239]
Jackson ignores the field to find the `JsonFormat` annotation, hence the default `LocalDateTime` deserializer using the ISO8601 format is used when the object is built via a non annotated constructor.

For reference I had a similar issue as https://github.com/FasterXML/jackson-module-kotlin/issues/118 in a pure Java project with a type annotated with lombok and the field was `final`, the type is only used for deserialization. 

```java
@Data
public class A {
    @JsonFormat(shape = STRING, pattern = DATE_TIME_OUTPUT_FORMAT)
    private final LocalDateTime date;

    // other fields
}
```

The actual delomboked code looks like : 

```java
public class A {
    @JsonFormat(shape = STRING, pattern = DATE_TIME_OUTPUT_FORMAT)
    private final LocalDateTime date;
    @Generated
    public A(LocalDateTime date, ...) {
        this.date = date;
    }

    public LocalDateTime getDate() {return this.date;}
    ...
}
```

This is kind of confusing, removing the `final` keyword and lombok generates a class that looks like :

```java
public class A {
    @JsonFormat(shape = STRING, pattern = DATE_TIME_OUTPUT_FORMAT)
    private final LocalDateTime date;

    public LocalDateTime getDate() {return this.date;}
    public void setDate(LocalDateTime date) {this.date = date; }
    ...
}
```

As you can see the annotation is still on the field, not on setters.
Even using the lombok setting `lombok.copyableAnnotations += com.fasterxml.jackson.annotation.JsonFormat` in `lombok.config` do not work as expected, yet for a different reason, the constructor is not found.

```java
public class A {
    @JsonFormat(
        shape = Shape.STRING,
        pattern = "dd/MM/yyyy HH:mm:ss"
    )
    private final LocalDateTime date;

    @Generated
    public A(@JsonFormat(shape = Shape.STRING,pattern = "dd/MM/yyyy HH:mm:ss") final LocalDateTime date, ...) {
        this.date = date;
    }

    @JsonFormat(
        shape = Shape.STRING,
        pattern = "dd/MM/yyyy HH:mm:ss"
    )
    @Generated
    public LocalDateTime getDate() {
        return this.date;
    }

    ...
}
```

In short I would have expected Jackson to check the field annotation even when the object is build with a constructor when the object is deserialized, without any workaround.

----------

Using jackson 2.8.11, unfortunately I didn't check with 2.9.x because dependencies are failing with the newer version.

**Comment by cowtowncoder** [comment]
I think you are close to the reason of problem, but I think I can provide bit more insight into details.
Jackson always looks into field annotations, too, as long as names match. The problem here, I think, is that constructor may be detected as "delegating" constructor, and not as "properties-based" one; and in former case, the whole input value is matched to the single constructor argument, and that argument is not considered to have name to match.
Canonical example would be something like:

```java
public class StringWrapper {
 private String s;

 @JsonCreator(mode = Creator.Mode.DELEGATING) 
 public StringWrapper(String s) { this.s = s; }
}
```

and would match JSON String value., like

```json
"Foobar"
```

But same approach works for other kinds of things, including dates serialized as Strings, or event POJOs.

But properties-based creator is different:

```java
public class NameObject {
   private String name;

  @JsonCreator(mode = JsonCreator.Mode.PROPERTIES)
  public NameObject(@JsonProperty("name") String name) {
     this.name = name;
  }
}
```

and this would match value like:

```json
{ "name" : "Foobar" }
```

Now: problem comes if `mode` is not declared AND there is just one parameter: which mode should be assumed? Heuristics used can get complicated, especially if constructor parameter name discovery is enabled (for Java 8+ and later).
So I am guessing that in your case, delegating mode was selected, and if so, annotation on field would not match: incoming Object has no name, field has.

Above is based on my guess of content you try to bind; if you include simple reproduction with input test data it's easy to confirm or deny.

**Comment by bric3** [comment]
Sorry I forgot to mention that indeed I'm using Java 8 and set the compiler with `-parameters`.
Actually as you mentionned my problem _was_ that the generated constructor is not annotated by `@JsonCreator` so the `mode` is not even set.

Also sorry second time in a row, my snippet shows a single field and a single parameter constructor, however this was to have a focused snippet, but the actual class is composed of multiple fields (so is the constructor)

-------------------

In the meantime I got my object working using the following lombok usage and config:

```java
@Data
@AllArgsConstructor(onConstructor = @__(@JsonCreator))
public class A {
    @JsonFormat(shape = STRING, pattern = DATE_TIME_OUTPUT_FORMAT)
    private final LocalDateTime date;
    // other fields
}
```

```properties
lombok.addLombokGeneratedAnnotation = true
lombok.copyableAnnotations += com.fasterxml.jackson.annotation.JsonFormat
```

As you may notice, the generated constructor is getting annotated with `@JsonCreator` without specify the `mode` hence `DEFAULT`. But I'm not sure why it is needed since there's only one constructor whose parameters have the same name as fields.


------

> Now: problem comes if mode is not declared AND there is just one parameter: which mode should be assumed? Heuristics used can get complicated, especially if constructor parameter name discovery is enabled (for Java 8+ and later).
> So I am guessing that in your case, delegating mode was selected, and if so, annotation on field would not match: incoming Object has no name, field has.

Hmm that's interesting, that's why I could not get some single field types to work.

**Comment by cowtowncoder** [comment]
Only one constructor should ever be annotated with `@JsonCreator`: Jackson does not try to figure out which one to use dynamically from multiple, but (should) throw(s) an exception indicating ambiguous definition. It is legal to define 1 constructor, 1 (static) factory method however, although only one of them will ever be used if so (I'd have to check source to remember which one has precedence).

But if there are more than 1 parameter to creator method (construct / static factory method), mode can only be `properties`: so that should only be needed for 1-parameter creator methods.

**Comment by gdabski** [comment]
I think this is the same problem that I'm experiencing. In my case during deserialization via constructor Jackson only considers `@JsonDeserialize(as=...)` if the constructor is marked as `@JsonCreator`. If it's not annotated as such, deserialization still proceeds (based on the fact there is only one constructor at all), but the property class refinement specified on the field is not taken into account.

I traced down difference in control flow to [`POJOPropertiesCollector:498-501`](https://github.com/FasterXML/jackson-databind/blob/179bb15c1e1d331df6c3690307546f46cba00d5f/src/main/java/com/fasterxml/jackson/databind/introspect/POJOPropertiesCollector.java#L498) where an explicit declaration of `@JsonCreator` is sought. When not present, the property is not linked to a constructor parameter in  [`POJOPropertiesCollector:515`](https://github.com/FasterXML/jackson-databind/blob/179bb15c1e1d331df6c3690307546f46cba00d5f/src/main/java/com/fasterxml/jackson/databind/introspect/POJOPropertiesCollector.java#L515). Then the clause in [`BeanDeserializerFactory:554`](https://github.com/FasterXML/jackson-databind/blob/179bb15c1e1d331df6c3690307546f46cba00d5f/src/main/java/com/fasterxml/jackson/databind/deser/BeanDeserializerFactory.java#L554) isn't triggered and a regular `FieldProperty` is added to the `BeanDeserializerBuilder` instead of a `CreatorProperty`. The right type of field deserializer is actually resolved based on the `@JsonDeserialize` annotation in [`BeanDeserializerBase:473`](https://github.com/FasterXML/jackson-databind/blob/179bb15c1e1d331df6c3690307546f46cba00d5f/src/main/java/com/fasterxml/jackson/databind/deser/BeanDeserializerBase.java#L473) but the subsequent attempt to replace the property in `creatorProps` collection fails inside [`BeanDeserializerBase:476`](https://github.com/FasterXML/jackson-databind/blob/179bb15c1e1d331df6c3690307546f46cba00d5f/src/main/java/com/fasterxml/jackson/databind/deser/BeanDeserializerBase.java#L476.).

This test is using a 2-property bean to avoid confusion surrounding the properties vs delegating mode.

```java
public class JsonCreatorJsonDeserializeAllInOneTest {

    public enum Suite {CLUBS, DIAMONDS, HEARTS, SPADES}

    public static class JdBean {

        private final String name;

        @JsonDeserialize(as = EnumSet.class)
        private final Set<Suite> suites;

        public JdBean(String name, Set<Suite> suites) {
            this.name = name;
            this.suites = suites;
        }

        // getters, equals, hashcode

    }

    public static class JdCreatorBean {

        private final String name;

        @JsonDeserialize(as = EnumSet.class)
        private final Set<Suite> suites;

        @JsonCreator
        public JdCreatorBean(String name, Set<Suite> suites) {
            this.name = name;
            this.suites = suites;
        }

        // getters, equals, hashcode

    }

    private final ObjectMapper mapper = new ObjectMapper();

    {
        mapper.registerModule(new ParameterNamesModule(Mode.PROPERTIES));
    }

    @Test
    public void shouldUseJsonDeserializeType_when_implicitCreatorPresent() throws JsonProcessingException {
        // given
        JdBean bean = new JdBean("Steven", EnumSet.of(SPADES, HEARTS));

        // when
        String serialized = mapper.writeValueAsString(bean);
        JdBean deserialized = mapper.readValue(serialized, JdBean.class);

        // then
        assertEquals(bean, deserialized);
        assertThat(deserialized.getSuites(), instanceOf(EnumSet.class));
    }

    @Test
    public void shouldUseJsonDeserializeType_when_explicitCreatorPresent() throws JsonProcessingException {
        // given
        JdCreatorBean bean = new JdCreatorBean("Steven", EnumSet.of(SPADES, HEARTS));

        // when
        String serialized = mapper.writeValueAsString(bean);
        JdCreatorBean deserialized = mapper.readValue(serialized, JdCreatorBean.class);

        // then
        assertEquals(bean, deserialized);
        assertThat(deserialized.getSuites(), instanceOf(EnumSet.class));
    }

}

```

I'm using Jackson 2.10.1 with parameter names module and compiling with `--parameters` using OpenJDK11.

**Comment by cowtowncoder** [comment]
@gdabski First of all, thank you for providing a test case!

It is possible this belongs to a class of problems that can not be solved for 2.x -- you are probably right that this is exactly due to problem of binding Creator-based and regular properties, in case of implicit creators (ones without `@JsonCreator` annotation). They are added at a bit later point and properties enclosed are not linked. From what I remember there is no easy way to make linkage work, and a full rewrite of property introspection is needed. I hope to do that  with 3.0 (`master`), but have not started that part yet.

Rewrite, if and when we get there, will require that introspection starts with finding and resolving Creators (as in, selecting which ones to use, not just candidates), fully, and only then including other accessors, linking all properties found.

Until then, explicit use of `@JsonCreator` annotation may be the only work-around.

**Comment by apple-corps** [comment]
It seems like it's confusing to use JsonFormat for the following.

https://www.reddit.com/r/javahelp/comments/10csjop/shaping_json_in_jackson_without_creating_an_object/

**Comment by cowtowncoder** [comment]
With 2.18 rewrite of (Creator) property discovery, this issue may be resolved.

If not, would need a reproduction. Closing for now, may be re-opened with full reproduction against 2.18.3.
